File: src/main/java/net/frozenorb/potpvp/util/TimeUtils.java
package net.frozenorb.potpvp.util;

import kotlin.jvm.JvmStatic;
import lombok.experimental.UtilityClass;
import lombok.val;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@UtilityClass
public final class TimeUtils {

    private static final SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy HH:mm");

    /**
     * Delegate to TimeUtils#formatIntoMMSS for backwards compat
     */
    public static String formatIntoHHMMSS(int secs) {
        return formatIntoMMSS(secs);
    }

    /**
     * Formats the time into a format of HH:MM:SS. Example: 3600 (1 hour) displays as '01:00:00'
     *
     * @param secs The input time, in seconds.
     * @return The HH:MM:SS formatted time.
     */
    public static String formatIntoMMSS(int secs) {
        // Calculate the seconds to display:
        int seconds = secs % 60;
        secs -= seconds;

        // Calculate the minutes:
        long minutesCount = secs / 60;
        long minutes = minutesCount % 60;
        minutesCount -= minutes;

        long hours = minutesCount / 60;
        return (hours > 0 ? (hours < 10 ? "0" : "") + hours + ":" : "") + (minutes < 10 ? "0" : "") + minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
    }

    public String formatLongIntoHHMMSS(Long secs) {
        int unconvertedSeconds = secs.intValue();
        return formatIntoMMSS(unconvertedSeconds);
    }

    /**
     * Formats time into a detailed format. Example: 600 seconds (10 minutes) displays as '10 minutes'
     *
     * @param secs The input time, in seconds.
     * @return The formatted time.
     */
    public static String formatIntoDetailedString(int secs) {
        if (secs == 0) {
            return "0 seconds";
        }
        int remainder = secs % 86400;

        int days = secs / 86400;
        int hours = remainder / 3600;
        int minutes = (remainder / 60) - (hours * 60);
        int seconds = (remainder % 3600) - (minutes * 60);

        String fDays = (days > 0 ? " " + days + " day" + (days > 1 ? "s" : "") : "");
        String fHours = (hours > 0 ? " " + hours + " hour" + (hours > 1 ? "s" : "") : "");
        String fMinutes = (minutes > 0 ? " " + minutes + " minute" + (minutes > 1 ? "s" : "") : "");
        String fSeconds = (seconds > 0 ? " " + seconds + " second" + (seconds > 1 ? "s" : "") : "");

        return ((fDays + fHours + fMinutes + fSeconds).trim());
    }

    /**
     * Formats time into a format of MM/dd/yyyy HH:mm.
     *
     * @param date The Date instance to format.
     * @return The formatted time.
     */
    public static String formatIntoCalendarString(Date date) {
        return (dateFormat.format(date));
    }

    /**
     * Parses a string, such as '1h4m25s' into a number of seconds.
     *
     * @param time The string to attempt to parse.
     * @return The number of seconds 'in' the given string.
     */
    public static int parseTime(String time) {

        if (time.equals("0") || time.equals("")) {
            return (0);
        }

        String[] lifeMatch = new String[]{"w", "d", "h", "m", "s"};
        int[] lifeInterval = new int[]{604800, 86400, 3600, 60, 1};
        int seconds = 0;

        for (int i = 0; i < lifeMatch.length; i++) {

            final Matcher matcher = Pattern.compile("([0-9]*)" + lifeMatch[i]).matcher(time);

            while (matcher.find()) {
                seconds += Integer.parseInt(matcher.group(1)) * lifeInterval[i];
            }

        }

        return (seconds);
    }

    /**
     * Gets the seconds between date A and date B. This will never return a negative number.
     *
     * @param a Date A
     * @param b Date B
     * @return The number of seconds between date A and date B.
     */
    public static int getSecondsBetween(Date a, Date b) {
        return (Math.abs((int) (a.getTime() - b.getTime()) / 1000));
    }

}

File: src/main/kotlin/net/frozenorb/potpvp/kt/util/TimeUtils.kt
package net.frozenorb.potpvp.kt.util

import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern

object TimeUtils {

    private val mmssBuilder: ThreadLocal<StringBuilder> = ThreadLocal.withInitial { java.lang.StringBuilder() }
    private val dateFormat: SimpleDateFormat = SimpleDateFormat("MM/dd/yyyy HH:mm")

    @JvmStatic
    fun formatIntoHHMMSS(secs: Int): String {
        return formatIntoMMSS(secs)
    }

    @JvmStatic
    fun formatLongIntoHHMMSS(secs: Long): String {
        val unconvertedSeconds = secs.toInt()
        return formatIntoMMSS(unconvertedSeconds)
    }

    @JvmStatic
    fun formatIntoMMSS(secs: Int): String {
        var secs = secs
        val seconds = secs % 60
        secs -= seconds
        var minutesCount = (secs / 60).toLong()
        val minutes = minutesCount % 60L
        minutesCount -= minutes
        val hours = minutesCount / 60L
        val result = mmssBuilder.get()
        result.setLength(0)
        if (hours > 0L) {
            if (hours < 10L) {
                result.append("0")
            }
            result.append(hours)
            result.append(":")
        }
        if (minutes < 10L) {
            result.append("0")
        }
        result.append(minutes)
        result.append(":")
        if (seconds < 10) {
            result.append("0")
        }
        result.append(seconds)
        return result.toString()
    }

    @JvmStatic
    fun formatLongIntoMMSS(secs: Long): String {
        val unconvertedSeconds = secs.toInt()
        return formatIntoMMSS(unconvertedSeconds)
    }

    @JvmStatic
    fun formatIntoDetailedString(secs: Int): String {
        if (secs == 0) {
            return "0 seconds"
        }
        val remainder = secs % 86400
        val days = secs / 86400
        val hours = remainder / 3600
        val minutes = remainder / 60 - hours * 60
        val seconds = remainder % 3600 - minutes * 60
        val fDays = if (days > 0) " " + days + " day" + if (days > 1) "s" else "" else ""
        val fHours = if (hours > 0) " " + hours + " hour" + if (hours > 1) "s" else "" else ""
        val fMinutes = if (minutes > 0) " " + minutes + " minute" + if (minutes > 1) "s" else "" else ""
        val fSeconds = if (seconds > 0) " " + seconds + " second" + if (seconds > 1) "s" else "" else ""
        return (fDays + fHours + fMinutes + fSeconds).trim { it <= ' ' }
    }

    @JvmStatic
    fun formatIntoShortString(secs: Int): String {
        if (secs == 0) {
            return "0 seconds"
        }

        val remainder = secs % 86400
        val days = secs / 86400
        val hours = remainder / 3600
        val minutes = remainder / 60 - hours * 60
        val seconds = remainder % 3600 - minutes * 60

        val builder = StringBuilder()

        if (days > 0) {
            if (hours > 0) {
                builder.append("$days day${if (days == 1) "" else "s"} $hours hour${if (hours == 1) "" else "s"}")
            } else {
                builder.append("$days day${if (days == 1) "" else "s"}")
            }
        } else if (hours > 0) {
            if (minutes > 0) {
                builder.append("$hours hour${if (hours == 1) "" else "s"} $minutes minute${if (minutes == 1) "" else "s"}")
            } else {
                builder.append("$hours hour${if (hours == 1) "" else "s"}")
            }
        } else if (minutes > 0) {
            if (seconds > 0) {
                builder.append("$minutes minute${if (minutes == 1) "" else "s"} $seconds second${if (seconds == 1) "" else "s"}")
            } else {
                builder.append("$minutes minute${if (minutes == 1) "" else "s"}")
            }
        }

        return builder.trim().toString()
    }

    @JvmStatic
    fun formatLongIntoDetailedString(secs: Long): String {
        val unconvertedSeconds = secs.toInt()
        return formatIntoDetailedString(unconvertedSeconds)
    }

    @JvmStatic
    fun formatIntoCalendarString(date: Date): String {
        return dateFormat.format(date)
    }

    @JvmStatic
    fun parseTime(time: String): Int {
        if (time == "0" || time == "") {
            return 0
        }
        val lifeMatch = arrayOf("y", "w", "d", "h", "m", "s")
        val lifeInterval = intArrayOf(31_536_000, 604800, 86400, 3600, 60, 1)
        var seconds = -1
        for (i in lifeMatch.indices) {
            val matcher = Pattern.compile("([0-9]+)" + lifeMatch[i]).matcher(time)
            while (matcher.find()) {
                if (seconds == -1) {
                    seconds = 0
                }
                seconds += Integer.parseInt(matcher.group(1)) * lifeInterval[i]
            }
        }
        if (seconds == -1) {
            throw IllegalArgumentException("Invalid time provided.")
        }
        return seconds
    }

    @JvmStatic
    fun parseTimeToLong(time: String): Long {
        val unconvertedSeconds = parseTime(time)
        return unconvertedSeconds.toLong()
    }

    @JvmStatic
    fun getSecondsBetween(a: Date, b: Date): Int {
        return getSecondsBetweenLong(a, b).toInt()
    }

    @JvmStatic
    fun getSecondsBetweenLong(a: Date, b: Date): Long {
        val diff = a.getTime() - b.getTime()
        val absDiff = Math.abs(diff)
        return absDiff / 1000L
    }

}